Blocks: Add automatic handling of focus for RichText component

gziolo opened the pull request:
## Description
Closes #6213.

Follow up for #5029 where we introduced new context API from React 16.3 to apply `isSelected` prop to a few components used with `BlockEdit`. 

This PR explores this idea further to make it possible to enforce that there can only be one `RichText` component focused at the same time. Historically it was manually enforced by using local state which stored which component is focused and prevented others to be selected at the same time.

I proposed the following
> Yes, multiple RichText is quite difficult to handle and maintain. I will look into it next week. Maybe it wouldn't be that hard to automatically handle active state in a way where only one RichText can be active per block at a time.

@jorgefilipecosta also proposed together with @jayshenk the following:
> I think we can have only one active RichText for the whole post. Maybe we can just use a component id and store id of RichText that is active, it would simplify the logic of some blocks.

I started small by using `BlockEdit` scoped solution to aovid dealing with the global store. I didn't see an immediate need to use `wp.data` to solve the issue we had here. We might want to revisit this later, but at the moment it seems like using the context API limits the number of operation React would have to perform where the focused element changes. I might be wrong here though. 


## How has this been tested?

Made sure it works without any changes for all blocks and for those updated in particular:
- `Pullquote`
- `Quote`
- `Text columns`

## Screenshots <!-- if applicable -->

![auto-focus rich text](https://user-images.githubusercontent.com/699132/39259153-fd5785e2-48b5-11e8-8a93-e8e788bbf5f3.gif)

## Types of changes
Refactoring,  no changes in the UX or UI.

## Next steps

We use a similar technique that uses `withState` to change focus for Gallery and Image blocks, but they are more complex so I excluded them from this PR. However we might look into exposing a general purpose helper that would allow to wrap components and inject this behavior to them.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

gziolo committed:
Blocks: Add automatic handling of focus for RichText component

gziolo committed:
Core blocks: Add automatic handling for focus for Quote and Pullquote blocks

youknowriad commented on blocks/rich-text/index.js:
```diff
@@ -888,17 +888,30 @@ RichText.defaultProps = {
+	withInstanceId,
+	withBlockEditContext( ( context, ownProps ) => {
+		// When explicitly set as not selected, do nothing.
+		if ( ownProps.isSelected === false ) {
```
Should we add a deprecation message about the `isSelected` prop in RichText components?

gziolo replied:
We might, but as soon as we remove all occurrences in core blocks. Otherwise, we would flood the console with the deprecation messages. We should also make sure there aren't any cases when this possibility would still make sense.

Shelob9 replied:
I love having focus handled automatically by default. I'd like to be able to optionally control it myself by setting that prop using my own logic.

youknowriad reviewed (commented):
I like it but there might be a small bug on initial mount.

Try this:

 - Insert a paragraph block
 - Using the autocompleter switch to the text columns block
 - Type and move the mouse
 - Notice the toolbar is not shown even if the text columns has the focus.

gziolo committed:
Blocks: Show controls for newly created block

gziolo commented:
@youknowriad, great catch. I think I found a solution which solves the issue you described: https://github.com/WordPress/gutenberg/pull/6419/commits/f0edd3cf9e0ff064fcaf05e7e675b209274a7061. Let me know if you have some ideas how to improve this code. Maybe there is a way to avoid introducing another method in the BlockEdit context, which doesn't require passing down boolean param.

gziolo commented:
Rebased with master.

aduth commented on blocks/rich-text/index.js:
```diff
@@ -878,17 +878,35 @@ RichText.defaultProps = {
 		return {
-			isBlockSelected: isSelected,
+			isSelected: context.isSelected && context.focusedElement === ownProps.instanceId,
+			onSetup: () => {
```
Nit: Why arrow function rather than function property shorthand?

```js
onSetup() {
	// ...
}
```

gziolo replied:
Good question, I don’t have an answer 😜 I’ll update.

gziolo replied:
Updated with https://github.com/WordPress/gutenberg/pull/6419/commits/6f4a1365976f38759446ae453f9d58cc246c9f36

aduth commented on blocks/rich-text/index.js:
```diff
@@ -878,17 +878,35 @@ RichText.defaultProps = {
 		return {
-			isBlockSelected: isSelected,
+			isSelected: context.isSelected && context.focusedElement === ownProps.instanceId,
+			onSetup: () => {
```
When does `onSetup` get called? When the user first focuses into the field? I would have thought it'd be at each's initialization, though it doesn't seem like we should want to set the focused element at all until the user actually focuses within (i.e. only `onFocus`).

gziolo replied:
OnSetup gets called when RichText gets mounted (initial render of the editor or when new block gets added). OnFocus is triggered only when a block gets focus but not on the initial render. That’s why we need to maintain two methods.

aduth replied:
But I don't understand why we need to set the focused element merely by fact the RichText has initialized.

aduth commented on blocks/rich-text/index.js:
```diff
@@ -878,17 +878,35 @@ RichText.defaultProps = {
 		return {
-			isBlockSelected: isSelected,
+			isSelected: context.isSelected && context.focusedElement === ownProps.instanceId,
+			onSetup: () => {
```
Is this going to incur a lot of unnecessary renders due to the new function references being passed each time (bypassing any pure component behavior)?

gziolo replied:
There are also multiple props present  that change frequently causing rerenders. I tried to add some optimizations to avoid creating new references, but it wasn’t helpful at all. I didn’t notice any improvements

aduth replied:
"If you can't beat them, join them" ? 😆

gziolo replied:
It would be great to take a closer look at all updates triggered by RichText and better understand how and why it works this way.

gziolo committed:
Blocks: Use method shorthand instead of arrow functions

gziolo commented:
I rebased with master again and also added a small change in https://github.com/WordPress/gutenberg/pull/6419/commits/6f4a1365976f38759446ae453f9d58cc246c9f36.

mcsf reviewed (approved):
Nice work.

gziolo committed:
Blocks: Use onFocus on div to set focused element

gziolo commented:
@aduth I think I found a way to simplify this implementation using `onFocus` attached to `div` as you suggested in one of the comments. It seems to work exactly the same but in addition we no longer create new functions on every re-render :)

gziolo commented on blocks/rich-text/index.js:
```diff
@@ -896,17 +906,30 @@ RichText.defaultProps = {
+		}
+		// When explicitly set as selected or has onFocus prop provided,
+		// use value stored in the context instead.
+		if ( ownProps.isSelected === true || ownProps.onFocus ) {
```
Yes, it makes sense 👍

gziolo commented on core-blocks/pullquote/index.js:
```diff
@@ -68,12 +67,9 @@ export const settings = {
-	edit: withState( {
-		editable: 'content',
-	} )( ( { attributes, setAttributes, isSelected, className, editable, setState } ) => {
+	edit: ( { attributes, setAttributes, isSelected, className } ) => {
```
Will do on Monday 👍

gziolo replied:
Also applied to `Quote`

gziolo committed:
Blocks: Address review feedback

gziolo committed:
Core Blocks: Remove obsolete `isSelected` usage from Spacer block
